core/src/poh_timing_reporter.rs

```diff
@@ -0,0 +1,184 @@
+use {
+    solana_metrics::{datapoint::PohTimingPoint, datapoint_info},
+    solana_sdk::clock::Slot,
+    std::{collections::HashMap, fmt},
+};
+
+/// A PohTimestamp records timing of the events during the processing of a slot
+/// by the validator
+#[derive(Debug, Clone, Copy, Default)]
+pub struct SlotPohTimestamp {
+    /// Slot start time from poh
+    pub start_time: u64,
+    /// Slot end time from poh
+    pub end_time: u64,
+    /// Last shred received time from block producer
+    pub full_time: u64,
+}
+
+/// Display trait
+impl fmt::Display for SlotPohTimestamp {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "SlotPohTimestamp: start={} end={} full={}",
+            self.start_time, self.end_time, self.full_time
+        )
+    }
+}
+
+impl SlotPohTimestamp {
+    /// Return true if the timing points of all events are received.
+    pub fn is_complete(&self) -> bool {
+        self.start_time != 0 && self.end_time != 0 && self.full_time != 0
+    }
+
+    /// Update with timing point
+    pub fn update(&mut self, timing_point: PohTimingPoint) {
+        match timing_point {
+            PohTimingPoint::PohSlotStart(ts) => self.start_time = ts,
+            PohTimingPoint::PohSlotEnd(ts) => self.end_time = ts,
+            PohTimingPoint::FullSlotReceived(ts) => self.full_time = ts,
+        }
+    }
+
+    /// Return the time difference from slot full to slot start
+    pub fn slot_full_to_start_time(&self) -> i64 {
+        (self.start_time as i64).saturating_sub(self.full_time as i64)
```

Comment:
I also think we should get rid of the saturating sub and return negative values. 

This is because there can be scenarios where `slot_full` happens *before* slot start, specifically if the validator is falling behind on replay.

For instance:

1. Slot full block 3
5. Replay block 3 begins (takes a while)
6. Slot full block 4, sends the full signal
7. Replay block 3 finishes
8. Reset poh to block 3, resets Poh to block 3, sends the start signal

A negative value thus represents how delayed the beginning of slot replay was from receiving the full block

Reply:
It is already returning negative value because they are casted to i64 before subtract. 
It seems that `clippy` want to use saturating sub.